File: app/src/main/java/me/ykrank/s1next/binding/LargeImageViewBindingAdapter.java
package me.ykrank.s1next.binding;

import android.content.Context;
import android.databinding.BindingAdapter;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.request.RequestOptions;
import com.shizhefei.view.largeimage.LargeImageView;

import java.io.File;

import me.ykrank.s1next.data.api.Api;
import me.ykrank.s1next.data.pref.DownloadPreferencesManager;
import me.ykrank.s1next.widget.glide.model.ForcePassUrl;
import me.ykrank.s1next.widget.glide.viewtarget.LargeImageViewTarget;

public final class LargeImageViewBindingAdapter {

    private LargeImageViewBindingAdapter() {
    }

    @BindingAdapter({"url", "thumbUrl", "manager"})
    public static void loadImage(LargeImageView largeImageView, String url, @Nullable String thumbUrl, DownloadPreferencesManager manager) {
        if (TextUtils.isEmpty(url)) {
            return;
        }
        Context context = largeImageView.getContext();
        RequestBuilder<File> builder = Glide.with(context)
                .download(new ForcePassUrl(url));
        //avatar signature
        if (manager != null && Api.isAvatarUrl(url)) {
            builder = builder.apply(new RequestOptions()
                    .signature(manager.getAvatarCacheInvalidationIntervalSignature()));
        }

        builder.into(new LargeImageViewTarget(largeImageView) {
            @Override
            public void onLoadFailed(@Nullable Drawable errorDrawable) {
                if (thumbUrl != null) {
                    loadImage(largeImageView, thumbUrl, null, manager);
                } else {
                    super.onLoadFailed(errorDrawable);
                }
            }
        });
    }
}

File: app/src/main/java/me/ykrank/s1next/binding/LargeImageViewBindingAdapter.kt
package me.ykrank.s1next.binding

import android.graphics.drawable.Drawable
import android.net.Uri
import androidx.core.net.toFile
import androidx.databinding.BindingAdapter
import com.bumptech.glide.Glide
import com.github.ykrank.androidtools.util.L
import com.github.ykrank.androidtools.util.isFile
import com.github.ykrank.androidtools.widget.glide.viewtarget.LargeImageViewTarget
import com.shizhefei.view.largeimage.LargeImageView
import com.shizhefei.view.largeimage.factory.FileBitmapDecoderFactory
import me.ykrank.s1next.data.pref.DownloadPreferencesManager
import me.ykrank.s1next.widget.image.ImageBiz
import me.ykrank.s1next.widget.image.image

object LargeImageViewBindingAdapter {
    @JvmStatic
    @BindingAdapter("url", "thumbUrl", "manager", "show")
    fun loadImage(
        largeImageView: LargeImageView,
        url: Uri?,
        thumbUrl: Uri?,
        manager: DownloadPreferencesManager,
        show: Boolean
    ) {
        if (!show || url == null) {
            return
        }
        if (url.isFile()) {
            try {
                largeImageView.setImage(FileBitmapDecoderFactory(url.toFile()))
            } catch (e: Exception) {
                L.e(e)
                if (thumbUrl != null) {
                    loadImage(largeImageView, thumbUrl, null, manager, true)
                }
            }
            return
        }

        val imageBiz = ImageBiz(manager)
        val builder = Glide.with(largeImageView)
            .downloadOnly()
            .image(imageBiz, url, forcePass = true)
        builder.into(object : LargeImageViewTarget(largeImageView) {
            override fun onLoadFailed(errorDrawable: Drawable?) {
                if (thumbUrl != null) {
                    loadImage(largeImageView, thumbUrl, null, manager, show)
                } else {
                    super.onLoadFailed(errorDrawable)
                }
            }
        })
    }
}
